src/composables/secureStorageRef.ts

```diff
@@ -0,0 +1,153 @@
+import { Ref, ref, watch } from 'vue';
+import { validateMnemonic } from '@aeternity/bip39';
+
+import type { StorageKey } from '@/types';
+import { decrypt, encrypt } from '@/utils';
+import { IS_MOBILE_APP, STORAGE_KEYS } from '@/constants';
+import { WalletStorage } from '@/lib/WalletStorage';
+import { useAccounts } from './accounts';
+import { type ICreateStorageRefOptions, useStorageRef } from './storageRef';
+
+interface ICreateSecureStorageRefOptions<T> extends
+    Omit<ICreateStorageRefOptions<T>, 'serializer' | 'enableSecureStorage' | 'onBackgroundSync'> {
+  encryptionDisabled?: boolean;
+}
+
+/**
+ * Create a secure storage ref that encrypts and decrypts the value using the password.
+ * Is a wrapper around the useStorageRef composable.
+ * On mobile, the encryption and decryption are handled by the storageRef composable
+ * because we are using the SecureMobileStorage.
+ */
+export function useSecureStorageRef<T = string | object | any[]>(
+  initialState: T,
+  storageKey: StorageKey,
+  options: ICreateSecureStorageRefOptions<T> = {
+    encryptionDisabled: IS_MOBILE_APP,
+  },
+) {
+  /**
+   * Checks if the value needs to be decrypted and decrypts it using the encryption key.
+   * On mobile, the value is already decrypted by the storageRef composable.
+   */
+  async function getDecryptedValue(val: any): Promise<T> {
+    if (options.encryptionDisabled || !val) {
+      return val;
+    }
+    const { getEncryptionKey } = useAccounts();
+    const encryptionKey = await getEncryptionKey();
+    if (encryptionKey) {
+      let decryptedValue;
+      try {
+        decryptedValue = await decrypt(encryptionKey, val);
+      } catch {
+        throw new Error('Failed to decrypt the value');
+      }
+      try {
+        return JSON.parse(decryptedValue);
+      } catch {
+        return decryptedValue as T;
+      }
+    }
+    throw new Error('Failed to read the value');
+  }
+
+  /**
+   * Checks if the value needs to be encrypted and encrypts it using the encryption key.
+   * On mobile, the value is already encrypted by the storageRef composable.
+   */
+  async function getEncryptedValue(val: any): Promise<T> {
+    if (options.encryptionDisabled || !val) {
+      return val;
+    }
+    const { getEncryptionKey } = useAccounts();
+    const encryptionKey = await getEncryptionKey();
+    if (encryptionKey) {
+      return encrypt(encryptionKey, val) as T;
+    }
+    throw new Error('Failed to write the value');
+  }
+
+  const decryptedState = ref(initialState) as Ref<T>;
+
+  /** Should always hold encrypted values (both on storage and state) */
+  const encryptedState = useStorageRef<T>(null as T, storageKey, {
```

Comment:
On mobile this state is not "encrypted" so I think this name is misleading. Or I'm missing something?

Reply:
You are correct, on mobile outside of `SecureMobileStorage` the state is always decrypted

Reply:
The logic was changed so my comment is not anymore applicable.